bpf/bpf_host.c

```diff
@@ -526,6 +532,23 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 			return DROP_INVALID;
 	}
 
+#if !defined(ENABLE_NODEPORT) && defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
```

Comment:
OK, I think we should ignore whether `ENABLE_NODEPORT` is set or not. We are interested only in a case when `bpf_host` is attached to a native device (i.e. `!from_host`). The following patch (untested; IPv4-only) should do the job:
```
diff --git a/bpf/bpf_host.c b/bpf/bpf_host.c
index 2feaa7053..88a2b9e95 100644
--- a/bpf/bpf_host.c
+++ b/bpf/bpf_host.c
@@ -478,6 +478,12 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
                        if (ret < 0)
                                return ret;
                }
+               /* Verifier workaround: modified ctx access. */
+               if (!revalidate_data(ctx, &data, &data_end, &ip4))
+                       return DROP_INVALID;
+       }
+#endif /* ENABLE_NODEPORT */
+
 #if defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
                /* Without bpf_redirect_neigh() helper, we cannot redirect a
                 * packet to a local endpoint in the direct routing mode, as
@@ -488,13 +494,9 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
                 * we bypass request and reply path in the host namespace and
                 * do not run into this issue.
                 */
-               skip_redirect = true;
+               if (!from_host)
+                       skip_redirect = true;
 #endif /* NO_REDIRECT && !ENABLE_REDIRECT_FAST */
-               /* Verifier workaround: modified ctx access. */
-               if (!revalidate_data(ctx, &data, &data_end, &ip4))
-                       return DROP_INVALID;
-       }
-#endif /* ENABLE_NODEPORT */

 #ifdef ENABLE_HOST_FIREWALL
        if (from_host) {
```